### PR TITLE
Label frames with container_id

### PR DESCRIPTION
**Pull request by r1viollet**
# What does this PR do?

Add a label with container_id information
We won't lookup short lived processes.

# Motivation

When viewing data from a host, we want to understand if our process is running within a container.
Refer to following issue: https://github.com/DataDog/ddprof/issues/220

# Additional Notes

This introduces a ProcessHdr object as an attempt to move some of the responsibilities to this object. 

I have a libdatadog PR open : https://github.com/DataDog/libdatadog/pull/169
Though I will not wait for the libdatadog code. Most of the code would be the same.

# How to test the change?

I added some unit tests that make sure we are able to read the container_id information.
An end to end test is still missing.

### CODE REVIEW

File: include/ddprof_process.hpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  uint64_t _sample_counter = {};
+private:
+  static DDRes read_cgroup_id(pid_t pid, std::string_view path_to_proc,
+                              CGroupId_t &cgroup);
+
+  pid_t _pid;
+  CGroupId_t _cgroup;
+};
+
+// todo ref counting
```
Comment (r1viollet) [inline include/ddprof_process.hpp:47]: to be discussed, do we need ref counting for so little elements ?

File: src/ddprof_process.cc [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    // new process, parse cgroup
+    auto pair = _process_map.try_emplace(pid, pid);
+    if (pair.second) {
+      it = pair.first;
+    } else {
+      return std::nullopt;
+    }
+  }
+  ++(it->second._sample_counter);
+  if(!force && (it->second._sample_counter) < k_nb_samples_container_id_lookup) {
```
Comment (r1viollet) [inline src/ddprof_process.cc:114]: This is something I'll want to talk about.

**Comment by r1viollet** [comment]
> This PR will add a custom slice dimension using the container-id: Example of user experience image

I slightly changed the labels. They now read: 
- None : not running within a container
- Unknown: We did not perform the lookup (as we only do so for long-lived pids)

File: docs/Build.md [hunk 3]
```diff
@@ -34,3 +34,12 @@ MkBuildDir Rel
 RelCMake ../
 make -j 4 .
 ```
+
+### Updating libdatadog
+
+Head over to the libdatadog, do your changes and copy the library back to your vendor directory.
+
+```bash
+cp ${workdir}/libdatadog/headers/include/datadog/common.h ${workdir}/ddprof/vendor_gcc_unknown-linux-2.35_Debug/libdatadog-v2.1.0/include/datadog/common.h
```
Comment (r1viollet) [inline docs/Build.md:43]: not really relevant to the PR.

Comment (sanchda) [reply docs/Build.md:43]: Still worth mentioning, and much appreciated.  Every time I do this it still feels like a hack 😓

File: include/unwind_output.hpp [hunk 4]
```diff
@@ -24,8 +25,10 @@ struct UnwindOutput {
   void clear() {
     locs.clear();
     is_incomplete = true;
+    container_id = "unknown";
```
Comment (sanchda) [inline include/unwind_output.hpp:28]: Ideally we'd use a default value that conforms to the right part of the Datadog BE.  Ultimately, I think the huge incremental value is going to be relating the sub-profiles to different things we harvest in the product.

That said, the immediate consumer is the profiling BE, and whenever we have a NULL context I think we have a challenge with upgrading it into a distinct type (for filtering/discoverability), so this is good.

File: src/ddprof_process.cc [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    return ddres_warn(DD_WHAT_CGROUP);
+  }
+  std::string line;
+  while (std::getline(cgroup_file, line)) {
+    container_id = container_id_from_line(line);
+    if (container_id) {
+      return {};
+    }
+  }
+  container_id = "none";
```
Comment (sanchda) [inline src/ddprof_process.cc:69]: Ah, interesting.  So our default is `"unknown"`, but failure to process this upgrades it to `"none"`.  I think there's value in having a semantic distinction and I think these two values are as good as any other, but in the future we may want to be more descriptive.

**Review by sanchda — APPROVED** [review]
Staggering.  This is a tremendous step in the right direction and major context that we were missing.  Very happy to see this in the code!

**Review by sanchda — DISMISSED** [review]
Staggering.  This is a tremendous step in the right direction and major context that we were missing.  Very happy to see this in the code!

File: include/unwind_output.hpp (hunk 4 above)
Comment (r1viollet) [reply include/unwind_output.hpp:28]: Yeah, we might need to revisit this. I'm making it at least more transparent in the next commit.

File: src/ddprof_process.cc (hunk 5 above)
Comment (r1viollet) [reply src/ddprof_process.cc]: Thanks for highlighting that this was unclear. I added comments.
`none` means that it is operating outside a container.

**Comment by r1viollet** [comment]
CI was flaky due to a metric push failing.

**Pull request merged**